### ISSUE TITLE
allow const keyword to evaluate when --nodejs --harmony is used

### ISSUE THREAD

**Issue by rebelwarrior** [jashkenas/coffeescript#3571]
Hi I'm using the examples from PragProg's "NodeJS done the right way".
Hit this problem they are using the --harmony flag to use ECMA6's features.
Figured out how to pass the flag to coffeescript using: `coffee --nodejs --harmony`
This then throws an error:

``` coffeescript
error: reserved word "const"
const fs = require('fs')
^
```

for this line of code:
`const fs = require('fs')`
however this one executes correctly:

``` coffeescript
eval('const fs = require("fs")')
```

I believe the coffeescript parser is stopping on the 'const' keyword irrespective of the --harmony flag being passed.

**Comment by pflannery** [comment]
> It'd be nice to have variables be immutable by default, only assigned upon their initialization, and a special form for immutable variables, like mut a = 42 or maybe different symbols for denoting declaration/initialization and assignment like b = 0; b := 1 if cond. But it's probably too late for that now.

Yeah probably too late, my impression is that `let = mut` and `const` tells the compiler that this variable will always stay the same type and gives a compiler confidence to fully optimize without having to constantly second guess its type identity. It will be interesting to see what performance we get by using const when ES6 (aka ES2015) is over the bridge. I'm not seeing an optimization boost in native V8 yet but more of a on-par performance between `let` and `const`.

I found this a pretty good write up for let and const - https://strongloop.com/strongblog/es6-variable-declarations/

**Comment by epidemian** [comment]
> [...] my impression is that let = mut and const tells the compiler that this variable will always stay the same type and gives a compiler confidence to fully optimize without having to constantly second guess its type identity.

I don't expect to see code using `const` running faster than the equivalent code using `var` on any JS engine. If a variable is assigned only once, then there's no need to "constantly second guess its type"; the engine can treat it the same as a constant.

**Comment by maximzavadskiy** [comment]
@epidemian  I have a counter example here, that I just discovered today:
`coffeescript
defaultData =
  nestedProp : 5

data = defaultData # By reading this, it seems quite fine and easy
data.nestedProp += 1 # But this will modify defaultData!
`

**Comment by lydell** [comment]
@maximzavadskiy `const` would't belp in your example.

**Comment by vendethiel** [comment]
What you're looking for is `Object.freeze`.

**Comment by samvv** [comment]
+1 for this

**Comment by mallinkbird** [comment]
+1 for this

**Comment by sanzaru** [comment]
+1

**Comment by vendethiel** [comment]
you realize there's even a button on github to "+1" now, right?

**Comment by samvv** [comment]
Didn't know that, thanks!

**Comment by hooddanielc** [comment]
I think `const` is great. Using the keyword `const` gives me confidence that the reference will never be reassigned. I use it for defensive programming style. I believe it prevented me from creating software with hard to debug issues. For example, lets say a developer needs to detect an users year of age before the user enters a pornographic site using `moment.js` library.

_coffeescript version_
```
age = 18
```

_javascript version with strict mode enabled_
```
const age = 18;
```

The developer decides to put a global variable in the window object for required age and references that to block underaged users. Now lets say 6 months later the developer implements functionality to fade _all_ submit buttons out on click.

_coffeescript version_ 
```
# how much time for animation
age = 100

$(document).ready ->
  $('button[type="submit"]').on 'click', (e) ->
    $(e.target).fadeOut age
```

_javascript version with strict mode enabled_
```
// how much time for animation
const age = 100

$(document).ready(function () {
  $('button[type="submit"]').on('click', function (e) {
    $(e.target).fadeOut(age);
  });
});
  
```

WOOPS, looks like the developer accidentally let all underaged user into a pornagraphic site because they were using coffeescript. However, the es6 javascript compiler warned the developer because const keyword was being used.

This is why I would like to see the `const` keyword in coffeescript.

**Comment by debrice** [comment]
@hooddanielc it's likely that a website config will be hosted on an object. For example, a gambling website could have different limitations per country:

```js
const config = {
    US: { minAge: 21 },
    FR: { minAge: 16 }
};
```

as explained above, using const here only prevents you from re-assign the variable, it wont prevent mutation.

```js
config.US.minAge = 12; // this would work just fine
```

Sadly, `const` is likely to be misinterpreted and give a false sense of security.

**Comment by akre54** [comment]
Yeah keep in mind any checks that are client-side only can be easily bypassed using the developer tools (or any other method). This is not just a coffeescript issue. If you have critical data / checks on your client side code they can and will be bypassed by someone who cares enough.

**Comment by algesten** [comment]
I don't understand why people keep giving this deep object example as some kind of counter argument to `const`. yeah I get it. it's not a true immutable. can we move on now?

Why must coffee prevent me from using it even for the very helpful case of avoiding accidental variable name clashes?!

**Comment by akre54** [comment]
Because coffeescript doesn't have block scoping. It views variable shadowing as a Bad Part of javascript. See Jeremy's many posts on the topic, there are valid reasons for this decision.

**Comment by algesten** [comment]
ok. it was decided not being able to distinguish between variable declaration and reassignment is a good thing and therefore we must also avoid const.

**Comment by akre54** [comment]
It doesn't have `var`, why would it have const? This has been discussed.

**Comment by dadleyy** [comment]
for what its worth, I wouldn't mind if variables in all caps became `const`-ed:

```coffeescript
WHITESPACE = /^[^\n\S]+/
```
becomes
```javascript
const WHITESPACE = /^[^\n\S]+/;
```

but I would not be inclined to see `const` get added to coffeescript itself.

**Comment by rebelwarrior** [comment]
But it does generate "Var" in the bare JS it generates. It does not generate "const" even when I would like to.

**Comment by epidemian** [comment]
> for what its worth, I wouldn't mind if variables in all caps became `const`-ed

Maybe instead of that the CS compiler could error when trying to re-assign capitalized variables (kinda like what Ruby does):

```coffeescript
MESSAGE = 'hello'
foo = -> 
  MESSAGE = 'bye' # Error: cannot reassign MESSAGE constant.
```

Same thing would apply to classes as they are usually declared using ThisNamingConvention.

**Comment by digitaldesigndj** [comment]
@dadleyy UPPERCASE is completely backwards incompatible (@epidemian O.o). I mean, I don't write *that* code, but it's a big world out there. I don't see the advantage of UPPERCASE over const. Wait, terseness... yummy yummy terseness... but wait... pinky abuse... (all torn up inside)

**Comment by carlsmith** [comment]
The idea that you can use `const` to "avoid accidental variable names clashes" doesn't make sense. Having to think about which vars actually vary, and then label the constants, and update the labels whenever the code changes, is a lot of effort to avoid accidentally using a name that's already visible in the current scope. You should already know which variables are visible, else you are not familiar enough with the code or the runtime to competently edit that code anyway. It seems like bad practice to use `const` defensively, especially as it doesn't prevent you from clobbering any of the variable variables. It's a pointless safety net, which is full of big holes. 

If I understood their motivation, ES6 was designed to allow new languages to be implemented more easily, and has features that mostly exist for that use case. Constants make more sense in that context, and shouldn't be used all over your codebase. IMO.

**Comment by rebelwarrior** [comment]
Really the problem is that const can't be used at all in coffeescript which means it is taking a hard opinionated stance that it is a bad part, nay a terrible part of JS. 
I don't think so. May not be a recommended part or an optimal one but by banning it you loose a feature of Coffeescript which is available in JS. 
It is stuff like this that has made Coffeescript loose relevance vs plain JS.

**Comment by ozjd** [comment]
Is there a good reason not to switch the 'var' ouput to 'const' and 'let' ?

Perhaps with a flag - eg, --es6-vars and change UPPERCASE to const? Thoughts?

**Comment by carlsmith** [comment]
The main reason is just a general aversion to compilation flags. They've always been rejected in the past. Still, they haven't been ruled out totally, and ES6 support will probably force the issue. It's difficult to see how CS classes can become compatible with JS classes without a breaking change, so CoffeeScript will almost certainly have to add compiler flags, or be replaced by a new dialect or language.

**Comment by carlsmith** [comment]
`-1` on forcing constants to be uppercase. Most functions will be constants, and we don't want every function to have a `LOUD_NAME`. We also wouldn't be able to distinguish between class names and the names of regular functions if they all have to be uppercase. We also use camelCase a lot in CoffeeScript, and would be forced to use underscores if we only have uppercase constant names.

A constant assignment operator has been discussed elsewhere, so you'd do something like:

~~~ coffee
i = 0
pi := 3.141
~~~

A better looking operator would be nice, but it's a much less radical change at least.

**Comment by ozjd** [comment]
I get what you're saying, but I still think it's a better option than what we have right now.

**Comment by bjmiller** [comment]
Can people please stop pushing this issue?  `var`, `let`, and `const` are all the same in transpiled code.  The purpose of `let` and `const` is to prevent human error when writing code directly in JS.  Their utility is almost completely obviated by CS.

**Comment by rebelwarrior** [comment]
Unless of course you use CS to write JS that will go into other ppls code. Then the lack of "let" "const" etc is a severe deficiency. 
I get it that's not where CS wants to go. 
I'm over it. I moved on to elm and JS6.

**Comment by GeoffreyBooth** [comment]
There was much debate about `let` and `const` as part of designing CoffeeScript 2, and it was decided that they should not be supported: http://coffeescript.org/v2/#unsupported-let-const

See also https://github.com/coffeescript6/discuss/issues/31 and https://github.com/coffeescript6/discuss/issues/35
